File: java/spring-boot/src/main/java/com/backendtemplate/services/ProductSeedService.java
package com.backendtemplate.services;

import com.backendtemplate.repositories.ProductSeedRepository;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Profile("!no-db")
public class ProductSeedService {
    private final ProductSeedRepository repository;

    public ProductSeedService(ProductSeedRepository repository) {
        this.repository = repository;
    }

    @Transactional(rollbackFor = Exception.class)
    public int seed(String productId, int stock) {
        if (stock < 0) {
            throw new IllegalArgumentException("Stock must be nonnegative");
        }
        return repository.seedIfAbsent(productId, stock);
    }
}

File: kotlin/spring-boot/src/main/kotlin/com/backendtemplate/services/ProductSeedService.kt
package com.backendtemplate.services

import com.backendtemplate.repositories.ProductSeedRepository
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Profile("!no-db")
class ProductSeedService(private val repository: ProductSeedRepository) {
    @Transactional(rollbackFor = [Exception::class])
    fun seed(productId: String, stock: Int): Int {
        require(stock >= 0) { "Stock must be nonnegative" }
        return repository.seedIfAbsent(productId, stock)
    }
}
